releases/stable-branches-and-release-cycle.md

```diff
@@ -0,0 +1,79 @@
+# Producing and supporting stable releases
+
+## Summary
+
+Today the Submariner project releases one minor version a month without capacity
+for stable branches or bugfix releases. In the following proposal we define a plan
+to make this possible starting with 0.8.
+
+## Proposal
+
+### Alignment of branch -> image releases
+
+We produce currently `<image>:devel` for every master push, along with `<image>:x.y.z` for
+the releases.
+
+Changing the `master` branch name to `devel` would make this homogeneous so our image releases
+are simply `<image>:<branch-or-tag>`.
+
+### Creating one stable branch for every minor release
+
+Approaching the release of every minor or major version a new stable branch will be forked
+off `devel` branch in the form of `release-x.y` for every existing repository.
+
+Any changes related to pinning shipyard base image will be done in that branch.
+
+At that point in time, any patches which belong to the release will need to be merged in master
+and backported to `release-x.y`. For this reason, the later we branch, the less extra work
+will be necessary to produce a stable release.  
+
+### Maintaining a stable branch
+
+Any patches that are necessary for fixing bugs in a stable branch will need to be merged
+first into master, and the later stable branches.
+
+`release-0.8`  <-- `release-0.9` <-- `devel`
+
+Patches could be simple cherry-picks, or sometimes they could need specific adaptations
+to the branch.
+
+When a patch is merged to a stable branch the images for such branch are updated, for example
+`quay.io/submariner/submariner:release-0.9`
+
+### How long is a stable branch maintained
+
+Initially stable branches will be maintained until the next stable release is released,
+this would change in the future if we introduce the concept of long term support releases.
+
+### Reducing the cadence of minor releases
+
+Based on the previous ideas the existing minor version cadence doesn't seem realistic
+for maintaining stable branches. We don't want fake stable branches where patches
+are never backported.
+
+With the new cadence after each sprint we would produce a milestone release,
+for example
+
+* `0.9-milestone-1-rc0`, ... , `0.9-milestone-1`
+* `0.9-milestone-2-rc0`, ... , `0.9-milestone-2`
+* `0.9-milestone-3-rc0`, ... , `0.9-milestone-3`
+* `0.9-milestone-4-rc0`, ... , `0.9-milestone-4`
+
+In this case we won't do pinning of shipyard base image, but we will release and update
+the admiral & shipyard `go.mod` references.
+
+After a number of milestones instead we would follow the usual pattern for a release:
+
+* `0.9.0-rc0`, ..., `0.9.0` along with it's stable branch `release-0.9` as described
+previously.
```

Comment:
And at GA of that release we create the branch right?

Reply:
Shouldn’t we create the branch when we start preparing the release? (_I.e._ pin dependencies etc.)

Reply:
So we branch off at `rc0` then? As long as we don't branch along the milestones I think it makes sense.

Reply:
Whatever you guys think is better.

I believe branching near `ga` removes cherry-pick work... but complicates other stuff, like the pinning in go-mod. We can try any of those ways and see how it goes.

Reply:
Right, I don’t think we should branch for the milestones, as long as they are just snapshots of `devel`. My idea is that any non-`devel`-related changes should be on a branch; so for example we tag Shipyard, then change Admiral to pin the Dapper image on a branch leading to rc0, etc.

Reply:
So probably `rc` should be the point at which we branch (which is not a stage during a milestone).

That's also why I suggested not to use `rc` terminology here - https://github.com/submariner-io/enhancements/pull/20/files#r545148130